dask/array/core.py

```diff
@@ -3743,22 +3863,104 @@ def from_delayed(value, shape, dtype=None, meta=None, name=None):
     >>> value = dask.delayed(np.ones)(5)
     >>> array = da.from_delayed(value, (5,), dtype=float)
     >>> array
-    dask.array<from-value, shape=(5,), dtype=float64, chunksize=(5,), chunktype=numpy.ndarray>
+    dask.array<from-delayed, shape=(5,), dtype=float64, chunksize=(5,), chunktype=numpy.ndarray>
     >>> array.compute()
     array([1., 1., 1., 1., 1.])
+
+    >>> orig = da.zeros((6, 4), chunks=2)
+    >>> delayeds = orig.to_delayed(optimize_graph=False)
+    >>> delayeds[0, 0] = dask.delayed(np.ones)((2, 2))
+    >>> delayeds[-1, 1] -= 1
+    >>> array = da.from_delayed(delayeds, like=orig)
+    >>> array.compute()
+    array([[ 1.,  1.,  0.,  0.],
+           [ 1.,  1.,  0.,  0.],
+           [ 0.,  0.,  0.,  0.],
+           [ 0.,  0.,  0.,  0.],
+           [ 0.,  0., -1., -1.],
+           [ 0.,  0., -1., -1.]])
+
+    See also
+    --------
+    Array.to_delayed
+    map_blocks
+    blockwise
     """
-    from dask.delayed import Delayed, delayed
-
-    if not isinstance(value, Delayed) and hasattr(value, "key"):
-        value = delayed(value)
-
-    name = name or "from-value-" + tokenize(value, shape, dtype, meta)
-    dsk = {(name,) + (0,) * len(shape): value.key}
-    chunks = tuple((d,) for d in shape)
-    # TODO: value._key may not be the name of the layer in value.dask
-    # This should be fixed after we build full expression graphs
-    graph = HighLevelGraph.from_collections(name, dsk, dependencies=[value])
-    return Array(graph, name, chunks, dtype=dtype, meta=meta)
+
+    if not isinstance(value, np.ndarray):
+        if shape is not None:
+            ndim = len(shape)
+        elif chunks is not None:
+            ndim = len(chunks)
+        elif like is not None:
+            ndim = like.ndim
+        else:
+            raise ValueError(
+                "One of `chunks`, `shape`, or `like` must be specified when giving a single delayed value."
+            )
+
+        value = np.array([value], dtype=object, ndmin=ndim)
+
+    if value.dtype.kind != "O":
+        raise TypeError(
+            "`from_delayed` expects a NumPy object array containing `dask.delayed` objects, "
+            f"not {value.dtype}"
+        )
+    if like is not None:
+        if not isinstance(like, Array):
+            raise TypeError(f"`like=` must be a dask Array, not {type(like)}")
+        if chunks is None:
+            chunks = like.chunks
+        if meta is None:
+            meta = like._meta
+
+    if chunks is None and shape is None:
+        if not allow_unknown_chunksizes:
+            raise ValueError(
+                "Chunksizes must be specified, either via `chunks=` or `like=`. "
+                "If the sizes of individual chunks are actually unkown, you can pass "
+                "`allow_unknown_chunksizes=True`, but some downstream operations may not work."
+            )
+        chunks = tuple((float("nan"),) * s for s in value.shape)
```

Comment:
Any reason you chose `float("nan")` instead of `np.nan`, which is used elsewhere for unknown chunk sizes? They both seem fine so far as `np.isnan` checks go.

Reply:
```suggestion
        chunks = tuple((np.nan,) * s for s in value.shape)
```
I just didn't realize we used `np.nan` more commonly, I'll switch to that.